cs-verifiable-encryption/README.md

```diff
@@ -0,0 +1,135 @@
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+*Some terminology*: The entity which can decrypt the ciphertext is called *auditor*. The entity demanding a proof (ZKP) is 
+the *verifier* and the entity creating the proof is called the *prover*.   
+The *prover* holds 1 or more credentials and creates proof using them. The *auditor* has a public key for encryption and both the *prover* and *verifier* have access to it. It is important that the public key should be unambiguously known to both the *prover* and *verifier*. One way to achieve that would be for the *auditor* to put it on a tamper evident database accessible to both the *prover* and *verifier*. A blockchain is an eligible option. Now when the verifier is only allowing for provisional anonymity (like in examples above), the prover while creating the ZKP, encrypts a pre-decided (between *prover* and *verifier*) attribute(s) from the credential(s) using the *auditor*'s public key. The verifier will now verify the ZKP and also use *auditor*'s public key to check that the pre-decided attribute(s) from the credential(s) used to create the ZKP have been encrypted and the *auditor* when asked can decrypt them.  
+**Note**: The prover should never encrypt his link secret for the auditor.
+The verifiable encryption follows the Sigma protocol approach where the prover first creates the ciphertext of the attributes and ciphertext of blindings for those attributes. These blindings should match the ones used in the overall proof. After generating/receiving the challenge, it computes the response. The verifier will then use challnge and response to reconstruct the ciphertext over blindings and check for equality. The reason to structure it this way is for composability such that the main proving protocol should be able to use it.  
+This verifiable encryption is meant to be treated as a predicate.  
+The next section will describe the API of this protocol.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This protocol has been implemented in this [PR](https://github.com/hyperledger/ursa/pull/40). The PR contains tests to demonstrate how the protocol works.
+
+1. The auditor can create his encryption keypair using `CSKeypair::new`. This method takes the maximum number of attributes that can be encrypted using the keypair. The following example creates a keypair whose public key can be used to encrypt at max 3 attributes. Trying to encrypt more will result in an error.  The `CSKeypair` is an struct of 2 fields, a public key `CSEncPubkey` and a private key `CSEncPrikey`.  
```

Comment:
Need some more background on the auditor's "encryption keypair". Is this created just-in-time, uniquely, for every interaction where VE is used? Or is this a long-lived keypair that an auditor can reuse in many interactions?

Reply:
The encryption is randomized so it appears unique but I believe it’s meant to be used for long term keys

Reply:
Long term. Specified in the section containing conceptual description.